smart-contracts/testdata/contracts/v1/checkpointing-2.wat

```diff
@@ -0,0 +1,259 @@
+    (call $assert_eq_64
+          (call $invoke_self_receive (local.get $entrypoint))
+          (i64.const 1103806595071)) ;; Corresponds to 0b10000000011111111111111111111111111111111.
+
+    ;; Check that the state was rolled back.
+    (call $assert_eq_64
+          (call $get_state)
+          (i64.const 112))
+       (i32.const 0))
+
+ ;; Queries the self balance, sets state to 113, and fails.
+ (func $receive_b (export "test.b") (param i64) (result i32)
+       ;; Get the self address and put it at memory position 0..16.
+       (call $get_receive_self_address (i32.const 0))
+
+       ;; Get the contract balance and assert that the invocation succeeded.
+       (call $assert_invoke_success
+            (call $host_invoke
+                (i32.const 3)   ;; Tag for the contract balance query
+                (i32.const 0)   ;; Offset in memory to start reading from.
+                (i32.const 16))) ;; The length of the parameter.
+
+       ;; Set state to 113.
+       (call $set_state (i64.const 113))
+
+       ;; Fail with error code -1.
+       (i32.const -1))
+
+ ;; Calls entrypoint 'd', sets state to 113, and fails.
+ (func $receive_c (export "test.c") (param i64) (result i32)
+       ;; Call the entrypoint b'd' == 100.
+       (call $assert_invoke_success
+             (call $invoke_self_receive (i32.const 100)))
+
+       ;; ;; Set state to 113
+       (call $set_state (i64.const 113))
+
+       ;; Fail with error code -1.
+       (i32.const -1))
+
+;; Receive method 'e'.
+;; This first writes to an entry at [].
+;; Then call the method specified in the parameter (as a single character)
+;; and ensure that the method did not modify any state.
+(func $receive_e (export "test.e") (param i64) (result i32)
+    (local $entry i64)
+    (local $entrypoint i32)
+    ;; Get the parameter, which is 1 byte, indicating which receive function to call below.
+    ;; And check that one byte was read.
+    (call $assert_eq
+        (call $host_get_parameter_section
+            (i32.const 0) ;; parameter index
+            (i32.const 0) ;; pointer in wasm memory
+            (i32.const 1) ;; length
+            (i32.const 0)) ;; offset
+        (i32.const 1))
+    ;; Save the parameter to $entrypoint.
+    (local.set $entrypoint (i32.load8_u (i32.const 0)))
+
+   (local.set $entry (call $state_lookup_entry (i32.const 500) (i32.const 0)))
+   (call $state_entry_write (local.get $entry) (i32.const 0) (i32.const 8) (i32.const 0))
+
+   ;; Call the $entrypoint and make sure the state was not modified.
+   (call $assert_eq_64
+          (call $invoke_self_receive (local.get $entrypoint))
+          ;; Corresponds to 0b0000000000000000000000001000000000000000000000000000000000000000
+          ;; indicating the state has not changed, and there is a return value at index 1.
+          (i64.const 1099511627776))
+     ;; and check that the entry we read above is still alive.
```

Comment:
The code doing the check is missing here.

Reply:
```suggestion
     ;; and check that the entry we read above is still alive.
     (call $assert_eq
        (call $state_entry_read (local.get $entry) (i32.const 0) (i32.const 8) (i32.const 0))
        (i32.const 8)
   )
```

Reply:
I actually didn't really care about this check, so I removed it (but clearly not the comment).

I reintroduced it back since it does not hurt.